cmd/microservice-sui-worker/main.go

```diff
@@ -0,0 +1,511 @@
+		TokenDecimals:  decimalPlaces,
+		PackageId:      underlyingPackageId,
+		IsFluid:        false,
+	}
+
+	sui_queue.DecoratedTransfers(func(decoratedTransfers []sui_queue.DecoratedTransfer) {
+		transfersWithFees := make([]worker_types.TransferWithFee, 0)
+
+		for _, transfer := range decoratedTransfers {
+			var (
+				event      = transfer.Event
+				checkpoint = transfer.Checkpoint
+				userAction = transfer.UserAction
+
+				transactionDigest = userAction.TransactionHash
+				application_      = userAction.Application
+				transactionHash   = userAction.TransactionHash
+				logIndex          = userAction.LogIndex
+				tokenDetails      = userAction.TokenDetails
+
+				packageId = event.PackageId
+
+				gasFee = userAction.AdjustedFee
+			)
+
+			if tokenDetails.TokenShortName != baseTokenName {
+				continue
+			}
+
+			application, err := suiApps.ParseApplicationName(application_)
+			if err != nil {
+				log.Fatal(func(k *log.Log) {
+					k.Format(
+						"Unable to parse Sui application name %v in transaction %v",
+						application_,
+						transactionDigest,
+					)
+					k.Payload = err
+				})
+			}
+
+			// process app (set fee and application in user action)
+			// AppData is not yet used on Sui
+			feeData, _, emission, err := suiApps.GetApplicationFee(transfer, *event, *application)
+
+			utilityDetails, _ := utilities[packageId]
+			utility := utilityDetails.Utility
+
+			var (
+				fee    = feeData.Fee
+				volume = feeData.Volume
+				// application will be set by sui-user-actions
+			)
+
+			// we set the decorator and if there's an app fee
+			// if this is a non-application transfer, it has already been tracked as a user action
+			if fee == nil {
+				log.App(func(k *log.Log) {
+					k.Format(
+						"Skipping an application transfer for transaction %#v and application %#v!",
+						transactionDigest,
+						application,
+					)
+				})
+
+				continue
+			}
+
+			decorator := &worker_types.SuiWorkerDecorator{
+				SuiAppFees:     emission,
+				UtilityName:    utility,
+				ApplicationFee: fee,
+				Volume:         volume,
+			}
+
+			transferWithFee := worker_types.TransferWithFee{
+				UserAction: userAction,
+				Event:      event,
+				Checkpoint: checkpoint,
+				Decorator:  decorator,
+			}
+			transfersWithFees = append(transfersWithFees, transferWithFee)
+
+			sender, recipient, err := suiApps.GetApplicationTransferParties(
+				transfer,
+				*application,
+			)
+
+			if err != nil {
+				log.Fatal(func(k *log.Log) {
+					k.Format(
+						"Failed to get the sender and receiver for an application transfer with hash %s!",
+						transactionHash,
+					)
+
+					k.Payload = err
+				})
+			}
+			indexCopy := new(misc.BigInt)
+			indexCopy.Set(&logIndex.Int)
+
+			// adjust volume then convert to a big int
+			// no loss of precision as there won't be more decimals than tokenDecimals
+			volume = volume.Mul(volume, decimalsRat)
+			volumeAdjusted_, _ := new(big.Int).SetString(volume.FloatString(0), 0)
```

Comment:
There's a better way to do this by cutting it in half after normalising the right side. Since we know this works let's do this as a nice to have in the future.